### PR TITLE
1st try 8/03

### PR DESCRIPTION
**Pull request by Watxesnavo**

### CODE REVIEW

File: src/main/java/mate/jdbc/dao/CarDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+import mate.jdbc.lib.Dao;
+import mate.jdbc.model.Car;
+import mate.jdbc.model.Driver;
+import mate.jdbc.model.Manufacturer;
+import mate.jdbc.util.ConnectionUtil;
+
+@Dao
+public class CarDaoImpl implements CarDao {
+    @Override
+    public Car create(Car car) {
```
Comment (Rommelua) [inline src/main/java/mate/jdbc/dao/CarDaoImpl.java:21]: what about drivers in this method?

File: src/main/java/mate/jdbc/dao/CarDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+            throw new DataProcessingException("Couldn't get car by id " + id, e);
+        }
+        if (car != null) {
+            car.setDrivers(getDriversForCar(id));
+        }
+        return Optional.ofNullable(car);
+    }
+
+    @Override
+    public List<Car> getAll() {
```
Comment (Rommelua) [inline src/main/java/mate/jdbc/dao/CarDaoImpl.java:64]: what about drivers in this method?

File: src/main/java/mate/jdbc/dao/CarDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+                cars.add(getCar(resultSet));
+            }
+            return cars;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Couldn't get a list of carss from carsDB.", e);
+        }
+    }
+
+    @Override
+    public Car update(Car car) {
```
Comment (Rommelua) [inline src/main/java/mate/jdbc/dao/CarDaoImpl.java:81]: what about drivers in this method?

File: src/main/java/mate/jdbc/service/CarServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+            setDriverToCarStatement.executeUpdate();
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't add driver" + driver + " to car " + car, e);
+        }
+    }
+
+    @Override
+    public void removeDriverFromCar(Driver driver, Car car) {
+        String deleteQuery = "DELETE FROM cars_drivers WHERE car_id = ? and driver_id = ?";
+        try (Connection connection = ConnectionUtil.getConnection();
```
Comment (Rommelua) [inline src/main/java/mate/jdbc/service/CarServiceImpl.java:65]: You can use only CarDao in the Service layer. Do not use any Connection to DB

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/mate/jdbc/Main.java [hunk 5]
```diff
@@ -2,6 +2,5 @@
 
 public class Main {
     public static void main(String[] args) {
-        // test your code here
```
Comment (Ivan95kos) [inline src/main/java/mate/jdbc/Main.java:5]: You should test each method here

File: src/main/java/mate/jdbc/dao/CarDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+        return new Car(id, model, manufacturer);
+    }
+
+    private void insertDriverToDriverDB(List<Driver> drivers) {
+        String query = "INSERT INTO drivers (id, name, license_number) VALUES(?, ?, ?)";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query,
+                        Statement.RETURN_GENERATED_KEYS)) {
+            for (Driver driver : drivers) {
+                statement.setLong(1, driver.getId());
```
Comment (Ivan95kos) [inline src/main/java/mate/jdbc/dao/CarDaoImpl.java:187]: The ID should be generated automatically
```suggestion
```

File: src/main/java/mate/jdbc/dao/CarDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+            throw new DataProcessingException("can't insert driver to drivers table", a);
+        }
+    }
+
+    private void insertDriversToCarDriversDB(Car car, List<Driver> drivers)
+            throws SQLException {
+        String query = "INSERT INTO cars_drivers (car_id, driver_id) VALUES(?, ?)";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query,
+                        Statement.RETURN_GENERATED_KEYS)) {
```
Comment (Ivan95kos) [inline src/main/java/mate/jdbc/dao/CarDaoImpl.java:202]: You don't use statement.getGeneratedKeys() in this method
```suggestion
                PreparedStatement statement = connection.prepareStatement(query)) {
```

**Review by Ivan95kos — CHANGES_REQUESTED** [review]
Do not write code in the master branch, create a separate branch for this
Create appropriate tables in the database
Test each method in the main method before submitting for review

File: src/main/java/mate/jdbc/dao/CarDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            statement.setLong(1, id);
+            return statement.executeUpdate() > 0;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Couldn't delete car with id " + id, e);
+        }
+    }
+
+    public List<Car> getAllByDriver(Long driverId) {
```
[the remark this answers is not in the document]
Comment (olekskov) [reply src/main/java/mate/jdbc/dao/CarDaoImpl.java:121]: annotation not added

File: src/main/java/mate/jdbc/Main.java (hunk 5 above)
Comment (olekskov) [reply src/main/java/mate/jdbc/Main.java:5]: comment not removed

File: src/main/java/mate/jdbc/dao/CarDaoImpl.java (hunk 6 above)
Comment (olekskov) [reply src/main/java/mate/jdbc/dao/CarDaoImpl.java:188]: not fixed

File: src/main/java/mate/jdbc/dao/CarDaoImpl.java (hunk 7 above)
Comment (olekskov) [reply src/main/java/mate/jdbc/dao/CarDaoImpl.java:203]: not fixed

File: src/main/java/mate/jdbc/dao/CarDaoImpl.java [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+    public Car create(Car car) {
+        String query = "INSERT INTO cars (model, manufacturer_id) VALUES(?, ?)";
+        List<Driver> drivers = car.getDrivers();
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement creationStatement = connection.prepareStatement(query,
+                        Statement.RETURN_GENERATED_KEYS)) {
+            creationStatement.setString(1, car.getModel());
+            creationStatement.setLong(2, car.getManufacturer().getId());
+            creationStatement.executeUpdate();
+            //insertDriverToDriverDB(drivers);
```
Comment (olekskov) [inline src/main/java/mate/jdbc/dao/CarDaoImpl.java:30]: commented code shouldn't be present

**Review by olekskov — CHANGES_REQUESTED** [review]
All comments from previous review are ignored

File: src/main/java/mate/jdbc/dao/CarDaoImpl.java [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+                car.setId(resultSet.getObject(1, Long.class));
+            }
+            car.setDrivers(getDriversForCar(car.getId()));
+            insertDriversToCarDriversDB(car, drivers);
+            return car;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Couldn't create "
+                    + car + ". ", e);
+        }
+    }
```
Comment (aaananas) [inline src/main/java/mate/jdbc/dao/CarDaoImpl.java:41]:
```suggestion
        } catch (SQLException e) {
            throw new DataProcessingException("Couldn't create "
                    + car + ". ", e);
        }
        car.setDrivers(getDriversForCar(car.getId()));
        insertDriversToCarDriversDB(car, drivers);
        return car;
    }
```
you are opening an another connection inside those methods

image

File: src/main/java/mate/jdbc/dao/CarDaoImpl.java [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+        try {
+            String name = resultSet.getString("name");
+            String licenseNumber = resultSet.getString("license_number");
+            Driver driver = new Driver(name, licenseNumber);
+            driver.setId(resultSet.getLong("id"));
+            return driver;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get driver from DB", e);
+        }
+    }
```
Comment (aaananas) [inline src/main/java/mate/jdbc/dao/CarDaoImpl.java:168]:
```suggestion
    private Driver getDriver(ResultSet resultSet) throws SQLException {
            String name = resultSet.getString("name");
            String licenseNumber = resultSet.getString("license_number");
            Driver driver = new Driver(name, licenseNumber);
            driver.setId(resultSet.getLong("id"));
            return driver;
    }
```

File: src/main/java/mate/jdbc/dao/CarDaoImpl.java [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+            ResultSet resultSet = getAllCarsStatement.executeQuery();
+            List<Car> cars = new ArrayList<>();
+            while (resultSet.next()) {
+                cars.add(getCar(resultSet));
+            }
+            return cars;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't find cars in DB by driver driverId "
+                    + driverId, e);
+        }
```
Comment (aaananas) [inline src/main/java/mate/jdbc/dao/CarDaoImpl.java:138]:
```suggestion
        } catch (SQLException e) {
            throw new DataProcessingException("Can't find cars in DB by driver driverId "
                    + driverId, e);
        }
        insert drivers
```

File: src/main/java/mate/jdbc/dao/CarDaoImpl.java [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+            for (Driver driver : driversOfCar) {
+                statement.setLong(3, driver.getId());
+                statement.executeUpdate();
+            }
+            return car;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Couldn't update "
+                    + car + " in carsDB.", e);
+        }
+    }
```
Comment (aaananas) [inline src/main/java/mate/jdbc/dao/CarDaoImpl.java:106]:
```suggestion
        } catch (SQLException e) {
            throw new DataProcessingException("Couldn't update "
                    + car + " in carsDB.", e);
        }
        delete relations from db
        insert new updated
    }
```

File: src/main/java/mate/jdbc/dao/CarDaoImpl.java [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+                PreparedStatement statement
+                        = connection.prepareStatement(query)) {
+            statement.setString(1, car.getModel());
+            statement.setLong(2, car.getManufacturer().getId());
+            statement.setLong(4, car.getId());
+            statement.setLong(5, car.getId());
+            for (Driver driver : driversOfCar) {
+                statement.setLong(3, driver.getId());
+                statement.executeUpdate();
+            }
```
Comment (aaananas) [inline src/main/java/mate/jdbc/dao/CarDaoImpl.java:100]:
```suggestion
            statement.setString(1, car.getModel());
            statement.setLong(2, car.getManufacturer().getId());
            statement.setLong(4, car.getId());
            statement.setLong(5, car.getId());
```

File: src/main/java/mate/jdbc/dao/CarDaoImpl.java [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                cars.add(getCar(resultSet));
+            }
+            return cars;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Couldn't get a list of cars from carsDB.", e);
+        }
```
Comment (aaananas) [inline src/main/java/mate/jdbc/dao/CarDaoImpl.java:80]:
```suggestion
        } catch (SQLException e) {
            throw new DataProcessingException("Couldn't get a list of cars from carsDB.", e);
        }
        insert drivers for each car
```

**Review by aaananas — CHANGES_REQUESTED** [review]

**Review by okuzan — CHANGES_REQUESTED** [review]
please fix all the previous comments

File: src/main/java/mate/jdbc/dao/CarDaoImpl.java [hunk 16]
```diff
@@ -0,0 +1,232 @@
...
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            ResultSet resultSet = statement.executeQuery();
+            while (resultSet.next()) {
+                cars.add(getCar(resultSet));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Couldn't get a list of cars from carsDB.", e);
+        }
+        for (Car car : cars) {
+            insertAllDrivers(car);
```
Comment (olekskov) [inline src/main/java/mate/jdbc/dao/CarDaoImpl.java:81]: It's a get method, so we're require to get info from db and prepare related model to provide representation of this data. We don't need to update info in DB in any way. So insertAllDrivers(car); shouldn't be here

**Review by olekskov — CHANGES_REQUESTED** [review]

File: src/main/java/mate/jdbc/dao/CarDaoImpl.java [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+            ResultSet resultSet = creationStatement.getGeneratedKeys();
+            if (resultSet.next()) {
+                car.setId(resultSet.getObject(1, Long.class));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Couldn't create "
+                    + car + ". ", e);
+        }
+        insertAllDrivers(car);
+        car.setDrivers(getDriversForCar(car.getId()));
```
Comment (uliana-andrushchuk) [inline src/main/java/mate/jdbc/dao/CarDaoImpl.java:39]: You don't need to set drivers here. This car already has drivers set

**Review by kozhukhovsky — APPROVED** [review]

**Pull request closed without merging**